File: lib/web_admin/utils/skill_scoring.dart
// Client-side skill score (mirrors backend skill_scoring.py).

double mediaQualityPercent(double averageQualityScore) {
  if (averageQualityScore <= 0) return 0;
  if (averageQualityScore <= 1) return averageQualityScore * 100;
  return averageQualityScore.clamp(0, 100);
}

double _pct(double num, double den, {double defaultValue = 0}) {
  if (den <= 0) return defaultValue;
  return (num / den * 100).clamp(0, 100);
}

double approvedUploadsPercent(int approved, int total) => _pct(approved.toDouble(), total.toDouble());

double taskParticipationPercent(int completed, int assigned) =>
    _pct(completed.toDouble(), assigned.toDouble());

double aiAuthenticityPercent(int humanVerified, int totalChecked) {
  if (totalChecked <= 0) return 100;
  return _pct(humanVerified.toDouble(), totalChecked.toDouble());
}

double computeSkillScore({
  required double mediaQualityScore,
  required double approvedUploadsScore,
  required double taskParticipationScore,
  required double adminEvaluationScore,
  required double aiAuthenticityScore,
}) {
  return mediaQualityScore * 0.40 +
      approvedUploadsScore * 0.25 +
      taskParticipationScore * 0.20 +
      adminEvaluationScore * 0.10 +
      aiAuthenticityScore * 0.05;
}

String skillLevelFromScore(double skillScore) {
  if (skillScore >= 90) return 'Expert';
  if (skillScore >= 75) return 'Advanced';
  if (skillScore >= 60) return 'Intermediate';
  return 'Beginner';
}

Map<String, dynamic> buildSkillPayload({
  required double averageQualityScore,
  required int approvedUploads,
  required int totalUploads,
  required int completedTasks,
  required int assignedTasks,
  required double adminEvaluationScore,
  required int humanVerifiedUploads,
  required int totalCheckedUploads,
}) {
  final mediaQ = mediaQualityPercent(averageQualityScore);
  final approvedQ = approvedUploadsPercent(approvedUploads, totalUploads);
  final taskQ = taskParticipationPercent(completedTasks, assignedTasks);
  final adminQ = adminEvaluationScore.clamp(0, 100).toDouble();
  final aiQ = aiAuthenticityPercent(humanVerifiedUploads, totalCheckedUploads);
  final skillScore = computeSkillScore(
    mediaQualityScore: mediaQ,
    approvedUploadsScore: approvedQ,
    taskParticipationScore: taskQ,
    adminEvaluationScore: adminQ,
    aiAuthenticityScore: aiQ,
  );
  return {
    'average_quality_score': mediaQ,
    'approved_uploads': approvedUploads,
    'total_uploads': totalUploads,
    'completed_tasks': completedTasks,
    'assigned_tasks': assignedTasks,
    'admin_evaluation_score': adminQ,
    'ai_authenticity_score': aiQ,
    'media_quality_score': mediaQ,
    'approved_uploads_score': approvedQ,
    'task_participation_score': taskQ,
    'skill_score': skillScore,
    'skill_level': skillLevelFromScore(skillScore),
  };
}

File: access_backend/app/services/skill_scoring.py
"""Automatic member skill score and tier from measurable performance."""

from __future__ import annotations


def _pct(numerator: float, denominator: float, *, default: float = 0.0) -> float:
    if denominator <= 0:
        return default
    return min(100.0, max(0.0, (numerator / denominator) * 100.0))


def media_quality_percent(average_quality_score: float) -> float:
    """Normalize average quality to 0–100 (supports 0–1 or 0–100 input)."""
    if average_quality_score <= 0:
        return 0.0
    if average_quality_score <= 1.0:
        return round(average_quality_score * 100.0, 2)
    return round(min(100.0, average_quality_score), 2)


def approved_uploads_percent(approved_uploads: int, total_uploads: int) -> float:
    return round(_pct(float(approved_uploads), float(total_uploads)), 2)


def task_participation_percent(completed_tasks: int, assigned_tasks: int) -> float:
    return round(_pct(float(completed_tasks), float(assigned_tasks)), 2)


def ai_authenticity_percent(human_verified: int, total_checked: int) -> float:
    """No scans yet → 100 (no violation recorded)."""
    if total_checked <= 0:
        return 100.0
    return round(_pct(float(human_verified), float(total_checked)), 2)


def compute_skill_score(
    *,
    media_quality_score: float,
    approved_uploads_score: float,
    task_participation_score: float,
    admin_evaluation_score: float,
    ai_authenticity_score: float,
) -> float:
    return round(
        media_quality_score * 0.40
        + approved_uploads_score * 0.25
        + task_participation_score * 0.20
        + admin_evaluation_score * 0.10
        + ai_authenticity_score * 0.05,
        2,
    )


def skill_level_from_score(skill_score: float) -> str:
    if skill_score >= 90:
        return "Expert"
    if skill_score >= 75:
        return "Advanced"
    if skill_score >= 60:
        return "Intermediate"
    return "Beginner"


def build_skill_payload(
    *,
    average_quality_score: float,
    approved_uploads: int,
    total_uploads: int,
    completed_tasks: int,
    assigned_tasks: int,
    admin_evaluation_score: float,
    human_verified_uploads: int,
    total_checked_uploads: int,
) -> dict:
    media_q = media_quality_percent(average_quality_score)
    approved_q = approved_uploads_percent(approved_uploads, total_uploads)
    task_q = task_participation_percent(completed_tasks, assigned_tasks)
    admin_q = round(min(100.0, max(0.0, admin_evaluation_score)), 2)
    ai_q = ai_authenticity_percent(human_verified_uploads, total_checked_uploads)
    skill_score = compute_skill_score(
        media_quality_score=media_q,
        approved_uploads_score=approved_q,
        task_participation_score=task_q,
        admin_evaluation_score=admin_q,
        ai_authenticity_score=ai_q,
    )
    return {
        "average_quality_score": media_q,
        "approved_uploads": approved_uploads,
        "total_uploads": total_uploads,
        "completed_tasks": completed_tasks,
        "assigned_tasks": assigned_tasks,
        "admin_evaluation_score": admin_q,
        "ai_authenticity_score": ai_q,
        "media_quality_score": media_q,
        "approved_uploads_score": approved_q,
        "task_participation_score": task_q,
        "skill_score": skill_score,
        "skill_level": skill_level_from_score(skill_score),
    }
